### PR TITLE
CASSANDRA-18336

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java [hunk 1]
```diff
@@ -124,12 +135,29 @@ public void teardown()
...
         DatabaseDescriptor.setDiskFailurePolicy(testPolicy);
-        JVMStabilityInspector.inspectThrowable(t);
-        Assert.assertEquals(expectJVMKilled, killerForTests.wasKilled());
-        Assert.assertEquals(expectJVMKilledQuiet, killerForTests.wasKilledQuietly());
-        if (!expectJVMKilled) {
+        try
+        {
+            JVMStabilityInspector.inspectThrowable(t);
+        }
+        catch (OutOfMemoryError e)
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java:142]: this is eventually thrown by `JVMStabilityInspector.inspectThrowable(t);` if you follow it deeply. We just need to make the difference between oom thrown as part of the test and oom thrown _while running a test_.

File: src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java [hunk 2]
```diff
@@ -96,6 +96,20 @@ public void handleFSError(FSError e)
...
 
+    private boolean isCausedByOutOfMemoryException(Throwable error)
+    {
+        for (Throwable t = error; t != null; t = t.getCause())
+        {
+            if (t instanceof OutOfMemoryError)
+                return true;
+            for (Throwable s : t.getSuppressed())
+                if (s instanceof OutOfMemoryError)
+                return true;
```
Comment (Maxwell-Guo) [inline src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java]: code format : the "return true;" should have a "tab"  before ?
`if (s instanceof OutOfMemoryError)
                return true;`

File: src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java [hunk 3]
```diff
@@ -81,7 +81,7 @@ public void handleFSError(FSError e)
 
                 // for both read and write errors mark the path as unwritable.
                 DisallowedDirectories.maybeMarkUnwritable(new File(e.path));
-                if (e instanceof FSReadError)
+                if (e instanceof FSReadError && !isCausedByOutOfMemoryException(e))
```
Comment (Maxwell-Guo) [inline src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java]: what about change the function to "isNotCausedByOutOfMemeoryException", so no "!" is needed before the function ,also the function logic should be rewrite ~~~~

File: src/java/org/apache/cassandra/utils/JVMStabilityInspector.java [hunk 4]
```diff
@@ -168,7 +168,7 @@ else if (t instanceof UnrecoverableIllegalStateException)
             inspectThrowable(t.getCause(), fn);
     }
 
-    private static final Set<String> FORCE_HEAP_OOM_IGNORE_SET = ImmutableSet.of("Java heap space", "GC Overhead limit exceeded");
+    private static final Set<String> FORCE_HEAP_OOM_IGNORE_SET = ImmutableSet.of("Java heap space", "GC Overhead limit exceeded", "Java heap space test");
```
Comment (Maxwell-Guo) [inline src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: I think "Java heap space test" may be not needed here, as FORCE_HEAP_OOM_IGNORE_SET is not used in the ut.

File: test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java [hunk 5]
```diff
@@ -79,22 +88,24 @@ public DiskFailurePolicyTest(DiskFailurePolicy testPolicy, boolean isStartUpInPr
...
+                             { die, false, new FSReadError(new IOException(), "blah"), false, true, false},
+                             { ignore, false, new FSReadError(new IOException(), "blah"), true, false, false},
+                             { stop, false, new FSReadError(new IOException(), "blah"), false, false, false},
+                             { stop_paranoid, false, new FSReadError(new IOException(), "blah"), false, false, false},
+                             { die, false, new CorruptSSTableException(new IOException(), "blah"), false, true, false},
+                             { ignore, false, new CorruptSSTableException(new IOException(), "blah"), true, false, false},
+                             { stop, false, new CorruptSSTableException(new IOException(), "blah"), true, false, false},
+                             { stop_paranoid, false, new CorruptSSTableException(new IOException(), "blah"), false, false, false},
+                             { best_effort, false, new FSReadError(new IOException(new OutOfMemoryError("Java heap space test")), "best_effort_oom"), true, false, false},
```
Comment (Maxwell-Guo) [inline test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java]: as "Java heap space test" removed in FORCE_HEAP_OOM_IGNORE_SET , so I think some other like "Java heap space" can be used here. I don't know if this is the right suggestion.

**Review by Maxwell-Guo — CHANGES_REQUESTED** [review]

File: src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java (hunk 2 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java]: thanks

File: src/java/org/apache/cassandra/utils/JVMStabilityInspector.java (hunk 4 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: No, this has to be there in order to skip the logic in the method where it is referenced.

File: test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java (hunk 5 above)
Comment (smiklosovic) [reply test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java]: see above

File: src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java (hunk 3 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/DefaultFSErrorHandler.java]: I rewrote the logic to be more generic.

File: src/java/org/apache/cassandra/utils/JVMStabilityInspector.java (hunk 4 above)
Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: what I mean is  such information("Java heap space test") will certainly not exist in the actual production environment, and the purpose of existence （"Java heap space test"） seems to be used only for ut testing. If it is purely used for testing, then why the information in ut does not directly use the existing exception message

**Review by Maxwell-Guo — CHANGES_REQUESTED** [review]

File: src/java/org/apache/cassandra/utils/JVMStabilityInspector.java (hunk 4 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: @Maxwell-Guo  What if _test itself_ throws OOM exception with a message "Java heap space"? How would you make the difference between legit OOM thrown during tests and OOM thrown as part of the test itself?

Comment (smiklosovic) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: @Maxwell-Guo If you have better solution, I am all ears.

Comment (smiklosovic) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: @Maxwell-Guo  I rewrote it so it is not there anymore.

Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: what about using System.setProperty?
at https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/utils/JVMStabilityInspector.java#L182

`        if (FORCE_HEAP_OOM_IGNORE_SET.contains(oom.getMessage()))
            return;
        if (System.getProperty("oom.ut.test.message") != null && System.getProperty("oom.ut.test.message").contains(oom.getMessage()))
            return;
        logger.error("Force heap space OutOfMemoryError in the presence of", oom);
        // Start to produce heap space OOM forcibly.
        List<long[]> ignored = new ArrayList<>();`

and at https://github.com/apache/cassandra/blob/trunk/test/unit/org/apache/cassandra/service/DiskFailurePolicyTest.java#L57

we can setting 
`  static
    {
        System.setProperty("oom.ut.test.message", "Java heap space test");
    }
`
others can be left as it is

Comment (smiklosovic) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: I think the current solution I just pushed is good enough. You probably posted the comment before checking it?

Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/utils/JVMStabilityInspector.java]: sorry @smiklosovic , the version when I commented was not the latest, so I have such a question. 
+1 on this, and thanks for this work.

**Pull request closed without merging**